### PR TITLE
JavaScript Long-Polling

### LINKED ISSUE #9: Use JavaScript for Long Polling
## Description
Currently, we keep the user's view of the queue up to date by using long polling, achieved with a `meta` tag.

https://github.com/SamLab17/HelpHours/blob/f9966602fab70e7ccf025fa3839763ad125df25d/helphours/templates/view.html#L5-L11

@Andrew-William-Smith suspects this may be causing the rare Chrome pre-fetch issue and that using JavaScript could avoid this
"optimization" by Chrome.

## Open Questions:

Should we do this long polling if the queue is closed _and_ empty?

### PR DESCRIPTION
**Pull request by SamLab17**
Implements long polling on the client side using long-polling.
Uses a new `/queue` route to get a JSON version of the queue and renders it on the front-end.

Update:
Also adds long polling to the "Join Queue" page so that we can keep the submit button up to date when the queue opens and closes.
It polls the `queue_status` route which returns either `OPEN` or `CLOSED`.

### CODE REVIEW

**Review by ninaahmed — APPROVED** [review]
LGTM!

File: helphours/static/scripts/join.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
+// How often we check if the queue is open
+var UPDATE_INTERVAL_SECONDS = 5;
```
Comment (Andrew-William-Smith) [inline helphours/static/scripts/join.js]: Shouldn't this be a constant instead of a variable?

```suggestion
const UPDATE_INTERVAL_SECONDS = 5;
```

Comment (SamLab17) [reply helphours/static/scripts/join.js]: Yeah you're right, halfway through writing the JavaScript I gave up on trying to support stuff before ES6 lol

File: helphours/static/scripts/queue.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
+var UPDATE_INTERVAL_SECONDS = 5;
```
Comment (Andrew-William-Smith) [inline helphours/static/scripts/queue.js]: Same here.
```suggestion
const UPDATE_INTERVAL_SECONDS = 5;
```

File: helphours/static/scripts/join.js (hunk 1 above)
Comment (Andrew-William-Smith) [reply helphours/static/scripts/join.js]: If it becomes that much of an issue, then we can always run this through Babel to compile down to ES5 anyway.  Given that we're using the `fetch` API, I think that going all-in on ES6 makes the most sense.

File: helphours/static/scripts/queue.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+                output += '</form></div>';
+            }
+            output += '</div>';
+        }
+    }
+    displayQueue(output);
+}
+
+function displayQueue(content) {
+    document.getElementById('queue').innerHTML = content;
```
Comment (Andrew-William-Smith) [inline helphours/static/scripts/queue.js]: You can keep this for the current commit, but in the future we need to find a better way to manage page contents than setting `innerHTML`.  `innerHTML` can lead to some substantial security issues: namely, if a student submits a script tag as their name, they could cause arbitrary contents to display on everyone's screens.  No changes suggested for this review, but I'm going to create an issue to audit use of `innerHTML` throughout the app.

File: helphours/templates/join.html [hunk 4]
```diff
@@ -30,7 +31,7 @@ <h2 class="form-title">Join the Queue</h2>
             {% endif %}
 
             <div>
-                {{ form.submit(class="form-submit") }}
+                    {{ form.submit(class="form-submit", id="submit-button") }}
```
Comment (Andrew-William-Smith) [inline helphours/templates/join.html]:
```suggestion
                {{ form.submit(class="form-submit", id="submit-button") }}
```
Why is this extra indented?

**Review by Andrew-William-Smith — APPROVED** [review]
Looks good to me!  See the comments for potential future issues, but we don't need to address them in this PR.

**Review by ClosingBracket — APPROVED** [review]

File: helphours/static/scripts/queue.js (hunk 3 above)
Comment (SamLab17) [reply helphours/static/scripts/queue.js]: Actually that seems like a pretty big issue, I redid the JavaScript in a new commit to not use `innerHTML` if you want to take a look at it.

**Review by Andrew-William-Smith — APPROVED** [review]
LGTM!

**Pull request merged**